<a href="https://colab.research.google.com/github/simsekergun/DATA601/blob/main/Project1/DATA601_Project1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## DATA 601 Project-1 (Spring 2022)
### Deadline: March 31, 11:59 pm
Please download these the csv files <br>
File-1: Maryland Statewide Vehicle Crashes:  <br>
https://opendata.maryland.gov/Public-Safety/Maryland-Statewide-Vehicle-Crashes/65du-s3qu

File-2: Person Details:  <br>
https://opendata.maryland.gov/Public-Safety/Maryland-Statewide-Vehicle-Crashes-Person-Details-/py4c-dicf

File-3: Vehicle Details:  <br>
https://opendata.maryland.gov/Public-Safety/Maryland-Statewide-Vehicle-Crashes-Vehicle-Details/mhft-5t5y

<br>

We will only use "Accident Year = 2020" data. So, please remove all the unncessary rows. Names these files as "crashes.csv", "persons.csv", and "vehicles.csv", respetively. I will test your notebook with the files I created, so if you have a different naming, your code won't work.

### Q1
From these three csv files, create a single dataframe which will include following columns only <br>
- from the "crashes.csv", accident data (acc_dat), accident time (acc_time), weather description (weather_desc),report no, county (county_desc), latittude, and longitude
- from the "persons.csv", gender, data of birth, and state code
- from the "vehicles.csv", vehicle year, vehicle make, speed limit
<br>

Save your dataframe as a new csv file called "md_crash_2020.csv".

Note: All these csv files have the 'REPORT_NO' column. You need to use that column to combine these datasets.

In [2]:
import pandas as pd
df=pd.read_csv("crashes.csv")
df


C:\Users\diksh\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


YEAR QUARTER      LIGHT_DESC  LIGHT_CODE     COUNTY_DESC  COUNTY_NO  \
0      2020      Q2        Daylight        1.00       Baltimore        3.0   
1      2020      Q4             NaN        6.02       Baltimore        3.0   
2      2020      Q2             NaN        6.02  Baltimore City       24.0   
3      2020      Q2        Daylight        1.00      Montgomery       15.0   
4      2020      Q4  Dark Lights On        3.00       Baltimore        3.0   
...     ...     ...             ...         ...             ...        ...   
94202  2020      Q1        Daylight        1.00       Baltimore        3.0   
94203  2020      Q1  Dark No Lights        4.00       Baltimore        3.0   
94204  2020      Q1        Daylight        1.00    Anne Arundel        2.0   
94205  2020      Q1        Daylight        1.00    Anne Arundel        2.0   
94206  2020      Q1        Daylight        1.00       Baltimore        3.0   

       MUNI_DESC  MUNI_CODE     JUNCTION_DESC  JUNCTION_CODE  ...  \
0            NaN        NaN               NaN            NaN  ...   
1            NaN        NaN               NaN            NaN  ...   
2            NaN        NaN  Non Intersection            1.0  ...   
3            NaN        NaN               NaN            NaN  ...   
4            NaN        NaN               NaN            NaN  ...   
...          ...        ...               ...            ...  ...   
94202        NaN        0.0      Intersection            2.0  ...   
94203        NaN        0.0  Non Intersection            1.0  ...   
94204        NaN        0.0    Not Applicable            0.0  ...   
94205        NaN        NaN               NaN            NaN  ...   
94206        NaN        0.0               NaN           88.0  ...   

      FEET_MILES_FLAG_DESC  FEET_MILES_FLAG DISTANCE_DIR_FLAG  REFERENCE_NO  \
0                      NaN              NaN               NaN           NaN   
1                      NaN              NaN               NaN           NaN   
2                    Miles                M                 N           NaN   
3                      NaN              NaN               NaN           NaN   
4                      NaN              NaN               NaN           NaN   
...                    ...              ...               ...           ...   
94202                 Feet                F                 E           0.0   
94203                 Feet                F                 S           0.0   
94204                 Feet                F                 E           0.0   
94205                  NaN              NaN               NaN           NaN   
94206                 Feet                F                 N         695.0   

       REFERENCE_TYPE_CODE REFERENCE_SUFFIX        REFERENCE_ROAD_NAME  \
0                      NaN              NaN                        NaN   
1                      NaN              NaN                        NaN   
2                      NaN              NaN                  NORTH AVE   
3                      NaN              NaN                        NaN   
4                      NaN              NaN                        NaN   
...                    ...              ...                        ...   
94202                   UU              NaN  ENT TO GOLDEN RING CENTER   
94203                   UU              NaN                      ALLEY   
94204                   UU              NaN           MILE MARKER 33.0   
94205                  NaN              NaN                        NaN   
94206                   IS              NaN              BALTO BELTWAY   

        LATITUDE  LONGITUDE                                    LOCATION  
0      39.277263 -76.503693             POINT (-76.5036932 39.27726285)  
1      39.323620 -76.503080    POINT (-76.503080116667 39.323620083333)  
2      39.311025 -76.616429    POINT (-76.616429453205 39.311024794431)  
3      39.140680 -77.193413    POINT (-77.193412729561 39.140680249069)  
4      39.316200 -76.738446    PO

In [3]:
df_crash=df[["ACC_DATE","ACC_TIME","WEATHER_DESC","REPORT_NO","COUNTY_DESC","LATITUDE","LONGITUDE"]]
df_crash

ACC_DATE  ACC_TIME    WEATHER_DESC     REPORT_NO     COUNTY_DESC  \
0      20200618  15:15:00  Not Applicable    AE58770052       Baltimore   
1      20201027  17:50:00             NaN    AE63330016       Baltimore   
2      20200430  06:39:00             NaN    ADJ984001J  Baltimore City   
3      20200504  09:46:00             NaN    EJ7878001D      Montgomery   
4      20201204  21:45:00         Raining    AE5739006G       Baltimore   
...         ...       ...             ...           ...             ...   
94202  20200303  14:45:00             NaN    AE5710009H       Baltimore   
94203  20200223  06:00:00             NaN    AE6171001M       Baltimore   
94204  20200207  10:02:00         Raining  MDTA1162001Q    Anne Arundel   
94205  20200322  15:53:00             NaN    AC24100005    Anne Arundel   
94206  20200114  12:34:00         Raining    AE5921006N       Baltimore   

        LATITUDE  LONGITUDE  
0      39.277263 -76.503693  
1      39.323620 -76.503080  
2      39.311025 -76.616429  
3      39.140680 -77.193413  
4      39.316200 -76.738446  
...          ...        ...  
94202  39.336339 -76.491263  
94203  39.253802 -76.522181  
94204  38.992001 -76.371489  
94205  38.929431 -76.560887  
94206  39.415545 -76.613480  

[94207 rows x 7 columns]

In [4]:
df=pd.read_csv("persons.csv")
df.columns

Index(['SEX_DESC', 'SEX_CODE', 'CONDITION_DESC', 'CONDITION_CODE',
       'INJ_SEVER_DESC', 'INJ_SEVER_CODE', 'REPORT_NO', 'OCC_SEAT_POS_DESC',
       'OCC_SEAT_POS_CODE', 'PED_VISIBLE_DESC', 'PED_VISIBLE_CODE',
       'PED_OBEY_DESC', 'PED_OBEY_CODE', 'PED_TYPE_DESC', 'PED_TYPE_CODE',
       'PED_LOCATION_CODE', 'MOVEMENT_DESC', 'MOVEMENT_CODE',
       'PERSON_TYPE_DESC', 'PERSON_TYPE', 'ALCOHOL_TEST_DESC',
       'ALCOHOL_TEST_CODE', 'ALCOHOL_TESTTYPE_DESC', 'ALCOHOL_TESTTYPE_CODE',
       'DRUG_TEST_DESC', 'DRUG_TEST_CODE', 'DRUG_TESTRESULT_DESC',
       'DRUG_TESTRESULT_CODE', 'BAC_CODE', 'FAULT_FLAG_DESC', 'FAULT_FLAG',
       'EQUIP_PROB_DESC', 'EQUIP_PROB_CODE', 'SAF_EQUIP_DESC',
       'SAF_EQUIP_CODE', 'EJECT_DESC', 'EJECT_CODE', 'DATE_OF_BIRTH',
       'PERSON_ID', 'LICENSE_STATE_CODE', 'CLASS', 'CDL_FLAG_DESC', 'CDL_FLAG',
       'VEHICLE_ID', 'EMS_UNIT_LABEL', 'AIRBAG_DEPLOYED', 'YEAR', 'Quarter'],
      dtype='object')

In [5]:
#from the "persons.csv", gender, data of birth, and state code
df_person=df[df['YEAR']==2020][["SEX_CODE","DATE_OF_BIRTH","LICENSE_STATE_CODE","REPORT_NO"]]
df_person["DATE_OF_BIRTH"]

0         20010626.0
1         20100128.0
2                NaN
3         20150131.0
4         19630828.0
             ...    
203273    19700322.0
203274    19740822.0
203275    19610609.0
203276    19570323.0
203277    19850331.0
Name: DATE_OF_BIRTH, Length: 203278, dtype: float64

In [6]:
df_person = df_person[df_person['DATE_OF_BIRTH'].notna()]
df_person
df_person['DOB'] = pd.to_datetime(df_person['DATE_OF_BIRTH'], format='%Y%m%d')
df_person

SEX_CODE  DATE_OF_BIRTH LICENSE_STATE_CODE    REPORT_NO        DOB
0             M     20010626.0                NaN   CB6032008S 2001-06-26
1             F     20100128.0                NaN   ADK1220015 2010-01-28
3             M     20150131.0                NaN   AE440100R7 2015-01-31
4             M     19630828.0                NaN   ADJ683000T 1963-08-28
6             F     19860212.0                NaN   ADK116001S 1986-02-12
...         ...            ...                ...          ...        ...
203273        M     19700322.0                NaN  MSP651500RX 1970-03-22
203274        F     19740822.0                NaN  MCP3217000W 1974-08-22
203275        M     19610609.0                 MD  MSP7291002W 1961-06-09
203276        M     19570323.0                 MD   AE59780033 1957-03-23
203277        M     19850331.0                 MD   AC23530024 1985-03-31

[184593 rows x 5 columns]

In [7]:
#from the "vehicles.csv", vehicle year, vehicle make, speed limit
df=pd.read_csv("vehicle.csv")
df.columns

Index(['HARM_EVENT_DESC', 'HARM_EVENT_CODE', 'CONTI_DIRECTION_DESC',
       'CONTI_DIRECTION_CODE', 'DAMAGE_DESC', 'DAMAGE_CODE', 'MOVEMENT_DESC',
       'MOVEMENT_CODE', 'VIN_NO', 'REPORT_NO', 'CV_BODY_TYPE_DESC',
       'CV_BODY_TYPE_CODE', 'VEH_YEAR', 'VEH_MAKE', 'VEH_MODEL',
       'COMMERCIAL_FLAG_DESC', 'COMMERCIAL_FLAG', 'HZM_NUM',
       'TOWED_AWAY_FLAG_DESC', 'TOWED_AWAY_FLAG', 'NUM_AXLES', 'GVW_DESC',
       'GVW_CODE', 'GOING_DIRECTION_DESC', 'GOING_DIRECTION_CODE',
       'BODY_TYPE_DESC', 'BODY_TYPE_CODE', 'DRIVERLESS_FLAG_DESC',
       'DRIVERLESS_FLAG', 'FIRE_FLAG_DESC', 'FIRE_FLAG', 'PARKED_FLAG_DESC',
       'PARKED_FLAG', 'SPEED_LIMIT', 'HIT_AND_RUN_FLAG_DESC',
       'HIT_AND_RUN_FLAG', 'HAZMAT_SPILL_FLAG_DESC', 'HAZMAT_SPILL_FLAG',
       'VEHICLE_ID', 'TOWED_VEHICLE_CONFIG_DESC', 'TOWED_VEHICLE_CONFIG_CODE',
       'AREA_DAMAGED_CODE_IMP1', 'AREA_DAMAGED_CODE1', 'AREA_DAMAGED_CODE2',
       'AREA_DAMAGED_CODE3', 'AREA_DAMAGED_CODE_MAIN_DESC',
       'AREA_DAMAGED_

In [8]:
df_vehicle=df[df['YEAR']==2020][["VEH_YEAR", "VEH_MAKE","SPEED_LIMIT","REPORT_NO"]]
df_vehicle

VEH_YEAR   VEH_MAKE  SPEED_LIMIT     REPORT_NO
0            NaN    UNKNOWN           25    AE5057004X
1         2018.0     TOYOTA           55   MSP7221001P
2         2019.0  CHEVROLET           30    DA37010040
3            NaN    UNKNOWN           35  MDTA1791002C
4         2002.0      HONDA           35    AE456100BL
...          ...        ...          ...           ...
172565    2002.0        GMC           25    ADG915004Q
172566    2017.0      CHEVY           55    ZW01540018
172567    1997.0       BUIC           25    ZG0963000G
172568    2000.0       AUDI           45   MSP5995007D
172569    2007.0       FORD            0    DE3112003N

[172570 rows x 4 columns]

In [9]:
print(df_crash.head(5))
print(df_person.head(5))
print(df_vehicle.head(5))

   ACC_DATE  ACC_TIME    WEATHER_DESC   REPORT_NO     COUNTY_DESC   LATITUDE  \
0  20200618  15:15:00  Not Applicable  AE58770052       Baltimore  39.277263   
1  20201027  17:50:00             NaN  AE63330016       Baltimore  39.323620   
2  20200430  06:39:00             NaN  ADJ984001J  Baltimore City  39.311025   
3  20200504  09:46:00             NaN  EJ7878001D      Montgomery  39.140680   
4  20201204  21:45:00         Raining  AE5739006G       Baltimore  39.316200   

   LONGITUDE  
0 -76.503693  
1 -76.503080  
2 -76.616429  
3 -77.193413  
4 -76.738446  
  SEX_CODE  DATE_OF_BIRTH LICENSE_STATE_CODE   REPORT_NO        DOB
0        M     20010626.0                NaN  CB6032008S 2001-06-26
1        F     20100128.0                NaN  ADK1220015 2010-01-28
3        M     20150131.0                NaN  AE440100R7 2015-01-31
4        M     19630828.0                NaN  ADJ683000T 1963-08-28
6        F     19860212.0                NaN  ADK116001S 1986-02-12
   VEH_YEAR   VEH_MAK

In [10]:
df_crash2020 = df_person.merge(df_crash,on='REPORT_NO').merge(df_vehicle,on='REPORT_NO')
print(df_crash2020)

       SEX_CODE  DATE_OF_BIRTH LICENSE_STATE_CODE    REPORT_NO        DOB  \
0             M     20010626.0                NaN   CB6032008S 2001-06-26   
1             M     20010626.0                NaN   CB6032008S 2001-06-26   
2             F     20090512.0                NaN   CB6032008S 2009-05-12   
3             F     20090512.0                NaN   CB6032008S 2009-05-12   
4             F     19770907.0                 MD   CB6032008S 1977-09-07   
...         ...            ...                ...          ...        ...   
374336        F     19620213.0                 MD   AE60670015 1962-02-13   
374337        F     19570313.0                 MD  MSP6310007D 1957-03-13   
374338        F     19680128.0                 MD   ZH0688000H 1968-01-28   
374339        M     19880114.0                 MD  MSP3053007W 1988-01-14   
374340        M     20030108.0                 MD   AE54040056 2003-01-08   

        ACC_DATE  ACC_TIME WEATHER_DESC COUNTY_DESC   LATITUDE  LONGITUDE  

### Q2
a. For each county, determine the average age of the drivers who were involved in these accidents  <br>
b. Then list the sorted averages (from youngest to oldest) <br>

In [11]:
import datetime as dt
import numpy as np
df2=df_crash2020[["DOB"]]
df2['current']=dt.datetime.today().strftime("%Y-%m-%d")
df2['currentdate'] = pd.to_datetime(df2['current'], format='%Y-%m-%d')
df2['Age']=df2['currentdate']-df2['DOB']
df2.dtypes
df2.Age = df2.Age.astype('str')
df2.dtypes
df2['Age']=df2.Age.str[0:-5]
df2.Age = df2.Age.astype('int')
df2['Age_in_years']=df2['Age']/365
print("Average age of drivers in accidents "+str(int(df2['Age_in_years'].mean())))
#li=df2['Age_in_years']
#print(" Sorted ages :")
#li=li.tolist()
#print(li)
#print(li.sort())

#print(df2)

#2b. pending

<ipython-input-11-0d727bed0466>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['current']=dt.datetime.today().strftime("%Y-%m-%d")
<ipython-input-11-0d727bed0466>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['currentdate'] = pd.to_datetime(df2['current'], format='%Y-%m-%d')
<ipython-input-11-0d727bed0466>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: h

Average age of drivers in accidents 39


### Q3
a. Determine the total number of crashes per "vehicle make". List the sorted values (e.g. Honda, 1234)<br>
b. Determine average age of the car involved in these accidents for each "vehicle make".  <br>

In [12]:
#df3=df_crash2020.groupby("VEH_MAKE")
df_new=df_crash2020[df_crash2020["VEH_MAKE"].notna()]
#df_crash2020[df_new]
df_vehicle=df_new["VEH_MAKE"].value_counts()
df5=pd.DataFrame(df_vehicle)
#print(df5)
#df5.reset_index(inplace=True)
#df5
list(df5.itertuples(index=True,name=None))

[('FORD', 45578),
 ('HONDA', 37861),
 ('TOYOTA', 36756),
 ('NISSAN', 25734),
 ('DODGE', 14249),
 ('CHEVY', 13597),
 ('CHEVROLET', 13131),
 ('HYUNDAI', 12905),
 ('JEEP', 10571),
 ('TOYT', 9774),
 ('UNKNOWN', 8853),
 ('CHEV', 8769),
 ('KIA', 8268),
 ('ACURA', 6816),
 ('GMC', 6243),
 ('HOND', 5727),
 ('BMW', 5327),
 ('LEXUS', 4875),
 ('NISS', 4860),
 ('MAZDA', 4814),
 ('HYUN', 4503),
 ('SUBARU', 3916),
 ('CHRYSLER', 3563),
 ('BUICK', 2513),
 ('MERCEDES', 2381),
 ('INFINITI', 2341),
 ('MERZ', 2093),
 ('AUDI', 2083),
 ('VOLKSWAGEN', 1967),
 ('VOLVO', 1957),
 ('DODG', 1764),
 ('VOLK', 1747),
 ('RAM', 1592),
 ('VOLKSWAGON', 1556),
 ('CHRY', 1528),
 ('INTERNATIONAL', 1509),
 ('FREIGHTLINER', 1496),
 ('FRHT', 1400),
 ('CADILLAC', 1372),
 ('MITSUBISHI', 1313),
 ('LINCOLN', 1289),
 ('SUBA', 1251),
 ('PONTIAC', 1171),
 ('MERCURY', 1040),
 ('ACUR', 990),
 ('INTL', 990),
 ('MAZD', 920),
 ('INFI', 885),
 ('SATURN', 864),
 ('CADI', 860),
 ('MITS', 850),
 ('INFINITY', 827),
 ('VW', 790),
 ('MACK', 746)

In [16]:
df_crash2020.dtypes
#df_crash2020['ACCIDENT_DATE']=

SEX_CODE                      object
DATE_OF_BIRTH                float64
LICENSE_STATE_CODE            object
REPORT_NO                     object
DOB                   datetime64[ns]
ACC_DATE                       int64
ACC_TIME                      object
WEATHER_DESC                  object
COUNTY_DESC                   object
LATITUDE                     float64
LONGITUDE                    float64
VEH_YEAR                     float64
VEH_MAKE                      object
SPEED_LIMIT                    int64
ACCIDENT_YEAR                  int32
dtype: object

In [14]:
#b. Determine average age of the car involved in these accidents for each "vehicle make".
l=  df_crash2020['ACC_DATE'].astype('str')
#df_crash2020['ACCIDENT_TYPE']=df_crash2020['ACC_DATE']
#df_crash2020.drop('ACCIDENT_TYPE',axis=1)
df_crash2020['ACCIDENT_YEAR']=l.str[:4]
#converting date column to int to perform calculations
df_crash2020['ACCIDENT_YEAR']=df_crash2020['ACCIDENT_YEAR'].astype('int')


In [15]:
#df_crash2020["CAR_AGE"]=df_crash2020["ACCIDENT_YEAR"]-df_crash2020["VEH_YEAR"]
df_crash2020.dtypes
df_crash2020["CAR_AGE"]=df_crash2020["ACCIDENT_YEAR"]-df_crash2020["VEH_YEAR"]
df_crash2020
d=df_crash2020.groupby('VEH_MAKE')['CAR_AGE'].mean()
#d
#d has mean of all cars
#verifying data by filtering 
print(d['KIA'])
fil=df_crash2020['VEH_MAKE']=='KIA'
fil
print(df_crash2020.loc[fil]['CAR_AGE'].mean())
#d=df_crash2020.loc[:,'VEH_MAKE'].groupby('VEH_MAKE')['CAR_AGE'].mean()
#d
print("Avg age of car")
print(d)
dk= df['VEH_MAKE'].apply(lambda x: count(x) if x.startswith("0"))

SyntaxError: invalid syntax (<ipython-input-15-1e9cc2fcb2bc>, line 17)

In [ ]:
#df_crash2020[df_crash2020["VEH_MAKE"].startswith("0")]

### Q4
a. Determine all the accidents that happened in a circle centered at (39.016149, -77.117942) with a radius of 5 km. Among them,  <br>
b. Determine the number of accidents where the speed limit was above 51 miles per hour. <br>
c. Determine the number of accidents where the speed limit was below 41 miles per hour.  <br>
<br>
Hint: Please see the "P1_Helper_Distance_From_Coordinates.ipynb" that can be found at https://github.com/simsekergun/DATA601/tree/main/Project1

In [ ]:
def haversine_distance(lat1, lon1):
   r = 6371
   lat2= 39.016149
   lon2= -77.117942
   phi1 = np.radians(lat1)
   phi2 = np.radians(lat2)
   delta_phi = np.radians(lat2-lat1)
   delta_lambda = np.radians(lon2-lon1)
   a = np.sin(delta_phi / 2)**2 + np.cos(phi1) * np.cos(phi2) *   np.sin(delta_lambda / 2)**2
   res = r * (2 * np.arctan2(np.sqrt(a), np.sqrt(1-a)))
   return np.round(res, 2)

In [ ]:
#df_crash2020
from math import radians, sin, cos, acos
import numpy as np
#df['distance'] = df['col1'].apply(complex_function)
df_crash2020['Distance'] = df_crash2020.apply(lambda x: haversine_distance(x['LATITUDE'], x['LONGITUDE']),axis=1)
df_crash2020[df_crash2020['Distance']<=5]

In [ ]:
# Determine the number of accidents where the speed limit was above 51 miles per hour.
df_crash2020[df_crash2020['SPEED_LIMIT']>51]

In [ ]:
#Determine the number of accidents where the speed limit was below 41 miles per hour.
df_crash2020[df_crash2020['SPEED_LIMIT']<41]

### Q5
Download "Licensed Drivers By Sex" data from  <br>
https://github.com/simsekergun/DATA601/tree/main/Project1

a. Determine the number of drivers, who got in a car crash, from non-DMV states as a function of gender and list them <br>
`St Male Female` <br>
`NY 1510  622` <br>
`PA 1251 1312` <br>
... <br>
Note: DMV states: DC, MD, and VA  <br>
 <br>
b. Normalize the previous list with number of male drivers and female drivers for each state. e.g. There are 1 million male drivers in NY and 2 million female drivers, then the normalized numbers should be <br>
<br>
`St    Male_Norm  Female_Norm` <br>
`NY    1.510e-3  0.311e-3` <br>



In [38]:
df_gender=pd.read_csv("Genders.csv")
DMV=['Washington','Maryland','Virginia']
df_gender
df_new_crash=df_crash2020[df_crash2020['LICENSE_STATE_CODE'].notna()]
df_new_crash.drop(df_new_crash[df_new_crash.SEX_CODE=='U'].index, inplace=True)
df10=df_new_crash.groupby('LICENSE_STATE_CODE')
df11=df10['SEX_CODE'].value_counts()
df11=pd.DataFrame(df11)
df11.rename(columns={"SEX_CODE":"COUNT"},inplace=True)
df11=df11.reset_index()
#df11=df11.reset_index()
df11
#removing DMV states
df11.drop(df11[(df11.LICENSE_STATE_CODE == 'DC') | (df11.LICENSE_STATE_CODE == 'MD') | (df11.LICENSE_STATE_CODE == 'VA' )].index, inplace=True)
df11=df11.pivot(index="LICENSE_STATE_CODE",columns="SEX_CODE",values="COUNT")
#df11.loc['NY']
df11
#b Normalize the previous list with number of male drivers and female drivers for each state. e.g. There are 1 million male drivers in NY and 2 million female drivers, then the normalized numbers should be
df12 = df11.div(df11.sum(axis=1), axis=0)
df12
#df.drop(df[(df.score < 50) & (df.score > 20)].index)
#df11.melt(id_vars=['SEX_CODE'])
#df_gender.STATE=="Iowa"
#df_gender[[df_gender['STATE']=='Washington']]
#df_new_one= df_crash2020.merge(df_gender,on='REPORT_NO').merge(df_vehicle,on='REPORT_NO')

C:\Users\diksh\Anaconda3\lib\site-packages\pandas\core\frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


SEX_CODE                   F         M
LICENSE_STATE_CODE                    
AB                  1.000000       NaN
AK                  0.395833  0.604167
AL                  0.379630  0.620370
AR                  0.383333  0.616667
AS                       NaN  1.000000
...                      ...       ...
WI                  0.305556  0.694444
WV                  0.369716  0.630284
WY                  0.250000  0.750000
XX                  0.227239  0.772761
YT                       NaN  1.000000

[67 rows x 2 columns]

### Q6
a. On which day of the week (Monday through Sunday) do the most of the crashes happen? <br>
b. On which hour of the day (00 - 24) do the most of the crashes happen?  <br>

In [39]:
#df_crash2020.dtypes
#df_crash2020['ACC_DATE'].to_datetime()
#print(df_crash2020['ACC_DATE'])
d12=pd.to_datetime(df_crash2020['ACC_DATE'],format="%Y%m%d")
#print(d12)
d13 = d12.dt.day_name()
d13
d14=pd.DataFrame(d13.value_counts())
d14.reset_index(inplace=True)
d14.rename(columns = {"index": "Day"}, inplace = True)
d14[d14['ACC_DATE']==d14['ACC_DATE'].max()]

#d14['ACC_DATE']==d14.max()
#d14[d14['ACC_DATE']==d14.max()]

Day  ACC_DATE
0  Friday     63876

In [40]:
#On which hour of the day (00 - 24) do the most of the crashes happen?
d15=pd.DataFrame(df_crash2020['ACC_TIME'].value_counts())
d15.reset_index(inplace=True)
d15.rename(columns = {"index": "Day"}, inplace = True)
d15[d15['ACC_TIME']==d15['ACC_TIME'].max()]


Day  ACC_TIME
0  16:00:00      2227

### Q7
Download "Maryland Average Precipitation and Snow 2020" data from  <br>
https://github.com/simsekergun/DATA601/tree/main/Project1

You need assume that it is a rainy/snowy day if the second column is bigger than 0.05 or if the third column is bigger than 1; otherwise take that day as a non-rainy/snowy day. <br>
a. What is the ratio of the number of accidents happened in rainy/snowy days to the number of accidents happened in non-rainy/snowy days? <br>
b. What does this ratio tell us?

In [42]:
import pandas as pd
df=pd.read_csv("Snow.csv", encoding= 'unicode_escape')
print(df.columns)
#cond = (df['Precip in.'] > 0.05) or (df['Snow in.'] > 1)
#print(df.loc[cond])
df.loc[df['Precip in.'] > 0.05, 'day_type'] = True 
df.loc[df['Snow in.'] > 1, 'day_type'] = True 
df['day_type'] = df['day_type'].fillna(False)
df['day_type'].value_counts()


Index(['Date', 'Precip in.', 'Snow in.'], dtype='object')


False    235
True     131
Name: day_type, dtype: int64

In [ ]:
#b. What does this ratio tell us?

#More accidents happened on non-rainy/snowy days


#Q8
Use matplotlib and/or seaborn and show your talent in data visualization. You are free to visualize anything you want regarding this dataset.

In [ ]:
df['day_type'].value_counts()